### ISSUE TITLE
[BUG] Install-AksHci - File download error: hash mismatch

### ISSUE THREAD

**Issue by raj-11** [Azure/aks-hci#79]
**Describe the bug**
Getting the error "File download error: hash mismatch." while deploying AKS on Azure Stack HCI using PowerShell.

**To Reproduce**
Following the below article to install AKS on HCI
https://docs.microsoft.com/en-us/azure-stack/aks-hci/setup-powershell

1. Execute the below command listed in step 5.
2. Install-AksHci
3. Get the below error
File download error: hash mismatch. Expected
KES8v7Ok2YPV9QWCSvQu9e19te7TL71M5R2c5YLZ4B8= and received
Nm+KqYR6RSFbA3zFid1RxQRSl0nj2yPYyxOXRXDqX30=
At C:\Program
Files\WindowsPowerShell\Modules\MSK8SDownloadAgent\1.0.0.209\sfsv1.psm1:459
char:13
+             throw "File download error: hash mismatch. Expected $($_. ...
+             ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : OperationStopped: (File download e...yPYyxOXR
   XDqX30=:String) [], RuntimeException
    + FullyQualifiedErrorId : File download error: hash mismatch. Expected
   KES8v7Ok2YPV9QWCSvQu9e19te7TL71M5R2c5YLZ4B8= and received Nm+KqYR6RSFbA
  3zFid1RxQRSl0nj2yPYyxOXRXDqX30=

**Expected behavior**
the AKS installation to complete successfully.

**Screenshots**
![aks-install-3](https://user-images.githubusercontent.com/48088365/111210507-3ccc3700-85c5-11eb-9b33-456380e44abe.jpg)

**Environment (please complete the following information):**
Azure Stack HCI
Version: 10.0.17784

Logs:
[agent-log-0.zip](https://github.com/Azure/aks-hci/files/6144016/agent-log-0.zip)

**Comment by BenjaminArmstrong** [comment]
@raj-11 - there were some issues that occurred yesterday with the Azure AD outage.  Can you confirm that you are still seeing this problem today?

**Comment by raj-11** [comment]
@BenjaminArmstrong , i had the same error first time i was installing AKS last week and couldnt uninstall or reinstall after the error hence did a clean install of Azure Stack HCI and had the same error over the weekend. so is a repeatable issue and not linked with AD outage.

**Comment by dineshgovindasamy** [comment]
@nwoodmsft FYI

**Comment by mkostersitz** [comment]
@raj-11 We have looked into this one and the provided logs. At this point it seems like a corrupted download. 
Are you deploying with a proxy server which inspects and modifies the downloaded image? If so please check with your network administrator. If the issue re-occurs please let us know.

**Comment by raj-11** [comment]
@mkostersitz There is no proxy server. That error re-occurred on a fresh install of Stack HCI so wasnt the case of corrupt downloaded file in the cache. I have since moved the server to DMZ with lesser firewall restriction. Wiped clean the server, reinstalled Stack HCI and had the hash mismatch error again.
![hci-error](https://user-images.githubusercontent.com/48088365/113875922-232aa380-97af-11eb-9d19-9b4fcf3e0405.jpg)

is there any known issue with the image download? 
is there an alternative set of powershell commands that can be used to retry the installation, as the status for AksHci gets set to "installed" and the only way to retry is to do a fresh install of Stack HCI.

**Comment by raj-11** [comment]
> @raj-11 We have looked into this one and the provided logs. At this point it seems like a corrupted download.
> Are you deploying with a proxy server which inspects and modifies the downloaded image? If so please check with your network administrator. If the issue re-occurs please let us know.

@mkostersitz There is no proxy server. That error re-occurred on a fresh install of Stack HCI so wasnt the case of corrupt downloaded file in the cache. I have since moved the server to DMZ with lesser firewall restriction. Wiped clean the server, reinstalled Stack HCI and had the hash mismatch error again.
![hci-error](https://user-images.githubusercontent.com/48088365/113879177-2bd0a900-97b2-11eb-919d-0c73d7ea98ee.jpg)


is there any known issue with the image download?
is there an alternative set of powershell commands that can be used to retry the installation, as the status for AksHci gets set to "installed" and the only way to retry is to do a fresh install of Stack HCI.

**Comment by mkostersitz** [comment]
I am re-opening this issue. 
@raj-11 there is no known issue with downloading or any other commands. You should be able to do an Uninstall-AksHci before starting over which should be sufficient. 
@nwoodmsft  do you have any suspicion what could trigger an invalid hash creation?

**Comment by raj-11** [comment]
@mkostersitz Uninstall-AksHci fails with below error. the status of AksHci shows as installed if i run the install-akshci command again.
![aks-uninstall-error](https://user-images.githubusercontent.com/48088365/113917727-73b6f680-97d9-11eb-956d-0e032256237f.jpg)

Logs attached.
[akshcilogs.zip](https://github.com/Azure/aks-hci/files/6273824/akshcilogs.zip)
[eventlogs.txt](https://github.com/Azure/aks-hci/files/6273825/eventlogs.txt)

**Comment by mkostersitz** [comment]
Thanks for the details @raj-11 can you try `Uninstall-Moc` and `Uninstall-Kva` and see if that cleans it up? I will be looking at the logs shortly

**Comment by raj-11** [comment]
> Thanks for the details @raj-11 can you try `Uninstall-Moc` and `Uninstall-Kva` and see if that cleans it up? I will be looking at the logs shortly

Thanks @mkostersitz . Uninstalling Moc and Kva did address the uninstall issue. Was able to run the Install-AksHci again but it failed with same 'hash mismatch' error.

**Comment by mkostersitz** [comment]
@raj-11 Unfortunately we are unable to determine from the logs what is going wrong here. We did try to reproduce the issue locally and were able to deploy the February version successfully on our end without any issue. At this point best course of action would be to go and download the latest version of the AKS-HCI PowerShell Module, remove the current module and then start over with the March release.

**Comment by raj-11** [comment]
@mik99git I had removed the old PS modules and coied the latest version of Aks-Hci modules from March. Now i get an error with empty vnet name while executing the command Install-AksHci.
![hci-error2](https://user-images.githubusercontent.com/48088365/114395990-ef77c100-9b94-11eb-833d-c5905a6e8c98.jpg)

on checking the value of name property, its empty.

While creating the akshci network settings, we pass the value for vnetName but there is no parameter for Name. but the vnet object retured has a name property and intall-akshci expects a value for name property (not vnetName). any suggestion on how to fix this?
$vnet = New-AksHciNetworkSetting -vnetName $vSwitchName

**Comment by abhilashaagarwala** [comment]
Can you try this on the AKS-HCI GA release and let us know if it persists? https://github.com/Azure/aks-hci/releases/tag/AKS-HCI-2105

**Comment by mkostersitz** [comment]
Closing. This does not reproduce in GA. If you find the issue is not fixed, please open a new issue with new logs.

**Comment by OEI-Cgray** [comment]
I know this is closed, but I cannot reinstall AKSHCI today due to this same error.

`PS C:\Temp> Install-Akshci



WARNING: Install-Moc failed.
Exception [GetRelease error returned by API call: run: Corrupted download: ExpectedSize=60633811 SavedSize=40422540]
Stacktrace [at Test-DownloadSDKResponse, C:\Program Files\WindowsPowerShell\Modules\DownloadSdk\1.0.13\DownloadSdk.psm1:
line 390
at Get-DownloadSdkRelease, C:\Program Files\WindowsPowerShell\Modules\DownloadSdk\1.0.13\DownloadSdk.psm1: line 173
at Get-MocReleaseContent, C:\Program Files\WindowsPowerShell\Modules\Moc\1.0.28\Moc.psm1: line 2381
at Get-MocRelease, C:\Program Files\WindowsPowerShell\Modules\Moc\1.0.28\Moc.psm1: line 2328
at Install-MocInternal, C:\Program Files\WindowsPowerShell\Modules\Moc\1.0.28\Moc.psm1: line 1202
at Install-Moc, C:\Program Files\WindowsPowerShell\Modules\Moc\1.0.28\Moc.psm1: line 221
at Install-AksHciInternal, C:\Program Files\WindowsPowerShell\Modules\AksHci\1.1.36\AksHci.psm1: line 4554
at Install-AksHci, C:\Program Files\WindowsPowerShell\Modules\AksHci\1.1.36\AksHci.psm1: line 804
at <ScriptBlock>, <No file>: line 1]
WARNING: Name resolution of ca-42b4e8c6-a674-4cd5-8539-bceccf3d9d86.obsidianent.com failed`

**Comment by mkostersitz** [comment]
@OEI-Cgray Hi Chris, the error you are getting is different from the original Hash mismatch error. 
Could you open a new issue please?
